src/Microsoft.ML.TimeSeries/SrCnnEntireTransformer.cs

```diff
@@ -0,0 +1,910 @@
+
+                if (!data.Schema.TryGetColumnIndex(inputCol.Source[1], out var valueColumnIndex))
+                    throw host.ExceptSchemaMismatch(nameof(data.Schema), "inputValue", inputCol.Source[1]);
+                if (data.Schema[valueColumnIndex].Type != NumberDataViewType.Double)
+                    throw host.ExceptSchemaMismatch(nameof(data.Schema), "inputTimestamp", inputCol.Source[1], "Double", data.Schema[valueColumnIndex].Type.ToString());
+
+                List<double> trainingData = new List<double>();
+                List<long> trainingTimestamp = new List<long>();
+                DateTime prevTimestamp = DateTime.MinValue;
+                using (var cursor = data.GetRowCursor(data.Schema))
+                {
+                    var getTimestamp = cursor.GetGetter<DateTime>(data.Schema[timestampColumnIndex]);
+                    var getVal = cursor.GetGetter<double>(data.Schema[valueColumnIndex]);
+                    DateTime timestamp = default(DateTime);
+                    double val = double.NaN;
+                    while (cursor.MoveNext())
+                    {
+                        getTimestamp(ref timestamp);
+                        getVal(ref val);
+
+                        if (timestamp > prevTimestamp && !double.IsNaN(val))
+                        {
+                            trainingTimestamp.Add(timestamp.Ticks);
+                            trainingData.Add(val);
+                            prevTimestamp = timestamp;
+                        }
+                    }
+                }
+
+                host.CheckIO(trainingTimestamp.Count >= MinBatchSize, "Input must have no less than 12 valid points to fit a model.");
+
+                var batchSize = (options.BatchSize == -1) ? trainingData.Count : options.BatchSize;
+                List<double[]> batchResult = new List<double[]>();
+                for (int i = 0; i * batchSize < trainingData.Count; ++i)
+                {
+                    var len = Math.Min(batchSize, trainingData.Count - i * batchSize);
+                    if (len >= MinBatchSize)
+                    {
+                        batchResult.AddRange(TrainCore(trainingData.GetRange(i * batchSize, len), options, length));
+                    }
+                    else
+                    {
+                        batchResult.AddRange(TrainCore(trainingData.GetRange(i * batchSize - batchSize + len, batchSize), options, length));
+                    }
+                }
+
+                anomalyDict = trainingTimestamp.Zip(batchResult, (k, v) => new { k, v }).ToDictionary(x => x.k, x => x.v);
+            }
+
+            private static List<double[]> TrainCore(List<double> values, Options option, int length)
+            {
+                List<double[]> results = values.Select(x => new double[length]).ToList();
```

Comment:
There is a lot of LINQ and ToList() here. Is that really necessary? I am worried about the potential number of allocatons